extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -89,28 +93,53 @@ public Signature boundSignature() {
     public Object execute(int idxInPartition,
                           WindowFrameState currentFrame,
                           List<? extends CollectExpression<Row, ?>> expressions,
+                          boolean ignoreNulls,
                           Input... args) {
-        final int offset;
-        if (args.length > 1) {
-            Object offsetValue = args[1].value();
-            if (offsetValue == null) {
-                return null;
+        if (offset == null) {
+            if (args.length > 1) {
+                Object offsetValue = args[1].value();
+                if (offsetValue == null) {
+                    return null;
+                } else {
+                    offset = ((Number) offsetValue).intValue();
+                }
             } else {
-                offset = ((Number) offsetValue).intValue();
+                offset = 1;
             }
-        } else {
-            offset = 1;
         }
+        return execute(idxInPartition, offset, currentFrame, expressions, ignoreNulls, args);
+    }
 
-        var lagRowCells = currentFrame
-            .getRowInPartitionAtIndexOrNull(offsetDirection.getTargetIndex(idxInPartition, offset));
+    private Object execute(int idxInPartition,
+                           int offset,
+                           WindowFrameState currentFrame,
+                           List<? extends CollectExpression<Row, ?>> expressions,
+                           boolean ignoreNulls,
+                           Input... args) {
+        int targetIndex = offsetDirection.getTargetIndex(idxInPartition, offset);
+        var lagRowCells = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
         if (lagRowCells != null) {
             var lagRow = new RowN(lagRowCells);
             for (CollectExpression<Row, ?> expression : expressions) {
                 expression.setNextRow(lagRow);
             }
-
-            return args[0].value();
+            var lagValue = args[0].value();
+            if (lagValue == null && ignoreNulls) {
+                if (cache.containsKey(targetIndex)) {
+                    return cache.get(targetIndex);
+                }
+                lagValue = execute(idxInPartition,
```

Comment:
@mfussenegger, please help review this implementation. I included a cache to speed up when a long chain of nulls is present. The cache will remember the indices and the results when the src data are nulls, so that for the next invocation the previous results can be re-used. 
The problem with this implementation is that for `lead` the call stack will grow(but not for `lag`).  ex) lead(x) on x = [1,2,null,null,null,...,null,3]
I am kind of stuck with this method as the iterative calls to `WindowFunction.execute` is externally controlled by `WindowFunctionBatchIterator.computeAndInjectResults`.  If I need to handle this, I think I should internalize the iteration but would like your suggestion before proceeding since it would take quite a while.

Reply:
Would it be possible to remember the last null or not null positions instead of caching every result?

```
[1, 2, null, null, 4, ...]`
    ▲              ▲
    │              │
    │             current idx
    │
    │
    last_not_null_idx = 1
```


Just to clarify the expected behavior of `lag` - it should skip `null` values if `IGNORE NULLS` is set?

So in the example above:

- `offset = 1` wouldn't return `null` but `2` ?
- `offset = 2` wouldn't return `null` but `1` ?

Reply:
Thanks @mfussenegger,
> 
    offset = 1 wouldn't return null but 2 ? Yes
    offset = 2 wouldn't return null but 1 ? Yes

I have realized my idea on lead/lag was wrong. I was thinking more of skipping nulls in the direction instead of ignoring nulls altogether. I will first come up with a right implementation without caching all not null positions